ibis/backends/dask/execution/util.py

```diff
@@ -436,3 +437,20 @@ def helper(
     df = df.set_index(col_name, sorted=True, divisions=divisions)
 
     return df
+
+
+def is_order_preserving(expr: ir.Expr) -> bool:
+    """Detects if the operation preserves ordering."""
+
+    def _is_order_preserving(op):
```

Comment:
You might be able to use `ibis.expr.lineage.traverse` here to handle the recursion and iteration.

Reply:
I actually borrowed this from https://github.com/ibis-project/ibis/blob/master/ibis/expr/analysis.py#L1192. I'll look into seeing if `traverse` is cleaner.